### ISSUE TITLE
CREO2URDF – Start investigating how to retrieve child/parent links from joints

### ISSUE THREAD

**Issue by Nicogene** [icub-tech-iit/creo2urdf#35]
Right now we are traversing the chain assuming that the prev and curr links are respectively the parent and child of the joint, assuming that there is only one kin chain and the parts are in order in the assembly.

We would like to retrieve this info from the axis, in order to be able to manage multiple kin chains and remove this assumption.

**Comment by Nicogene** [comment]
I started working on this branch https://github.com/icub-tech-iit/creo2urdf/tree/mvp2.5, for now it is still a WIP.

The idea is to somehow populate a map traversing the assembly, containing for each axis the child and the parent link name and then use it when create the urdf.

**Comment by Nicogene** [comment]
Here is the PR:

- #36

**Comment by Nicogene** [comment]
I had the chance to test the latest changes w/ the ergocub upperbody asm (see https://github.com/icub-tech-iit/creo2urdf/issues/33#issuecomment-1618205102) and right now here is the result:

![immagine](https://github.com/icub-tech-iit/creo2urdf/assets/19152494/bc57f870-6ccb-4dd0-aacb-486d8a53da40)


The strange things is that the upperarm are for some reason missing in the urdf, it is under investigation.

**Comment by Nicogene** [comment]
Printing the model to file seems to containe the upper arm links and linked joints:

```
Model: 
  Links: 
    [0] SIM_ECUB_ROOT_LINK
    [1] SIM_ECUB_TORSO_1
    [2] SIM_ECUB_TORSO_2
    [3] SIM_ECUB_CHEST
    [4] SIM_ECUB_L_SHOULDER_1
    [5] SIM_ECUB_R_SHOULDER_1
    [6] SIM_ECUB_L_SHOULDER_2
    [7] SIM_ECUB_R_SHOULDER_2
    [8] SIM_ECUB_R_SHOULDER_3
    [9] SIM_ECUB_L_SHOULDER_3
    [10] SIM_ECUB_L_UPPERARM    <-----
    [11] SIM_ECUB_HEAD_NECK_2
    [12] SIM_ECUB_HEAD_NECK_3
    [13] SIM_ECUB_HEAD
    [14] SIM_ECUB_REALSENSE
    [15] SIM_ECUB_R_UPPERARM  <-----
  Frames: 
  Joints: 
    [0] SIM_ECUB_CHEST--SIM_ECUB_L_SHOULDER_1 (dofs: 1) : SIM_ECUB_CHEST<-->SIM_ECUB_L_SHOULDER_1
    [1] SIM_ECUB_L_SHOULDER_1--SIM_ECUB_L_SHOULDER_2 (dofs: 1) : SIM_ECUB_L_SHOULDER_1<-->SIM_ECUB_L_SHOULDER_2
    [2] SIM_ECUB_L_SHOULDER_3--SIM_ECUB_L_UPPERARM (dofs: 1) : SIM_ECUB_L_SHOULDER_3<-->SIM_ECUB_L_UPPERARM   <-----
    [3] SIM_ECUB_CHEST--SIM_ECUB_HEAD_NECK_2 (dofs: 1) : SIM_ECUB_CHEST<-->SIM_ECUB_HEAD_NECK_2
    [4] SIM_ECUB_HEAD_NECK_2--SIM_ECUB_HEAD_NECK_3 (dofs: 1) : SIM_ECUB_HEAD_NECK_2<-->SIM_ECUB_HEAD_NECK_3
    [5] SIM_ECUB_HEAD_NECK_3--SIM_ECUB_HEAD (dofs: 1) : SIM_ECUB_HEAD_NECK_3<-->SIM_ECUB_HEAD
    [6] SIM_ECUB_HEAD--SIM_ECUB_REALSENSE (dofs: 1) : SIM_ECUB_HEAD<-->SIM_ECUB_REALSENSE
    [7] SIM_ECUB_CHEST--SIM_ECUB_R_SHOULDER_1 (dofs: 1) : SIM_ECUB_CHEST<-->SIM_ECUB_R_SHOULDER_1
    [8] SIM_ECUB_R_SHOULDER_1--SIM_ECUB_R_SHOULDER_2 (dofs: 1) : SIM_ECUB_R_SHOULDER_1<-->SIM_ECUB_R_SHOULDER_2
    [9] SIM_ECUB_R_SHOULDER_3--SIM_ECUB_R_UPPERARM (dofs: 1) : SIM_ECUB_R_SHOULDER_3<-->SIM_ECUB_R_UPPERARM  <-----
    [10] SIM_ECUB_TORSO_1--SIM_ECUB_TORSO_2 (dofs: 1) : SIM_ECUB_TORSO_1<-->SIM_ECUB_TORSO_2
    [11] SIM_ECUB_ROOT_LINK--SIM_ECUB_TORSO_1 (dofs: 1) : SIM_ECUB_ROOT_LINK<-->SIM_ECUB_TORSO_1
    [12] SIM_ECUB_TORSO_2--SIM_ECUB_CHEST (dofs: 1) : SIM_ECUB_TORSO_2<-->SIM_ECUB_CHEST
```


But then this is the resulting urdf:
```
<?xml version="1.0"?>
<robot name="ECUB_UPPERBODY">
  <link name="SIM_ECUB_ROOT_LINK">
    <inertial>
      <mass value="2.823751482824641"/>
      <origin xyz="0.0024748962432214403 -1.050794470993344e-4 0.023450368787955644" rpy="0 -0 0"/>
      <inertia ixx="0.004380033669258687" ixy="-1.4369039688385886e-6" ixz="1.4098441290530498e-5" iyy="0.005630606285299465" iyz="-2.3299340994588874e-5" izz="0.0036166810108083755"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_root_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_root_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_ROOT_LINK--SIM_ECUB_TORSO_1" type="revolute">
    <origin xyz="-2.23890229767886e-18 -5.551115123142428e-20 0.060500000000000095" rpy="0 -2.2470258116354473e-44 5.656623812198896e-30"/>
    <axis xyz="-1 5.656623812198896e-30 1.4239886692466155e-15"/>
    <parent link="SIM_ECUB_ROOT_LINK"/>
    <child link="SIM_ECUB_TORSO_1"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_TORSO_1">
    <inertial>
      <mass value="2.016853871356948"/>
      <origin xyz="-0.009639973621975028 6.168494736287859e-4 0.08244231288506343" rpy="0 -0 0"/>
      <inertia ixx="0.0070947444084796685" ixy="6.975203277646223e-6" ixz="-1.2586871224024975e-5" iyy="0.005816920576319008" iyz="7.227736063238349e-4" izz="0.002953654264837281"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_torso_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_torso_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_TORSO_1--SIM_ECUB_TORSO_2" type="revolute">
    <origin xyz="-7.30886630439206e-17 -4.930380657631324e-32 0.11800000000000001" rpy="3.9719461100401744e-15 2.710505431213761e-20 1.0587911275090697e-22"/>
    <axis xyz="2.1175823115833007e-22 1 -3.97151242917118e-15"/>
    <parent link="SIM_ECUB_TORSO_1"/>
    <child link="SIM_ECUB_TORSO_2"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_TORSO_2">
    <inertial>
      <mass value="0.7051627057068411"/>
      <origin xyz="0.0015724674776760397 0.0036223466761378916 0.026019504666210665" rpy="0 -0 0"/>
      <inertia ixx="6.675525890436374e-4" ixy="-4.198623927073284e-6" ixz="7.014302140884286e-5" iyy="0.0012590296877101022" iyz="5.398163583772658e-6" izz="0.0011458228109270142"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_torso_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_torso_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_TORSO_2--SIM_ECUB_CHEST" type="revolute">
    <origin xyz="7.32502235717738e-17 2.561326957877574e-16 0.0644699999999999" rpy="-3.971868545919581e-15 -4.385099568445115e-23 -3.1763738272196853e-22"/>
    <axis xyz="2.0848912754007678e-15 1.8581752484505413e-15 -1"/>
    <parent link="SIM_ECUB_TORSO_2"/>
    <child link="SIM_ECUB_CHEST"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_CHEST">
    <inertial>
      <mass value="8.065471632648594"/>
      <origin xyz="-0.010721051678060343 -3.482064053382885e-4 0.07611063517438789" rpy="0 -0 0"/>
      <inertia ixx="0.04493883052455447" ixy="-1.7611941682820536e-4" ixz="-1.2962048932620152e-4" iyy="0.04845475402461777" iyz="3.12408928056946e-4" izz="0.03068768729801644"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_chest.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_chest.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_CHEST--SIM_ECUB_L_SHOULDER_1" type="revolute">
    <origin xyz="-0.013518676842610159 0.06704761117227499 0.12351981080941767" rpy="0.2536123238732541 -0.05203160550449479 -0.27527873241387235"/>
    <axis xyz="-0.2714473275142867 0.9351131265310296 -0.22777135240129398"/>
    <parent link="SIM_ECUB_CHEST"/>
    <child link="SIM_ECUB_L_SHOULDER_1"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_L_SHOULDER_1">
    <inertial>
      <mass value="0.7913756606310043"/>
      <origin xyz="1.482427542201549e-4 0.06782971819041982 5.036196460998787e-4" rpy="0 -0 0"/>
      <inertia ixx="4.33126825498053e-4" ixy="2.049760682973134e-4" ixz="-1.9449541410974415e-4" iyy="0.0011548041680336772" iyz="5.017853886187208e-5" izz="0.0011461196178955738"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_l_shoulder_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_l_shoulder_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_CHEST--SIM_ECUB_HEAD_NECK_2" type="revolute">
    <origin xyz="0.004861512347971912 4.1253975310307907e-16 0.22529999999999994" rpy="-1.541978321110266e-15 -2.0480667158215426e-15 -1.389802281096855e-15"/>
    <axis xyz="-1.3900353108727565e-15 1 2.3071799250996495e-15"/>
    <parent link="SIM_ECUB_CHEST"/>
    <child link="SIM_ECUB_HEAD_NECK_2"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_HEAD_NECK_2">
    <inertial>
      <mass value="0.12100314943032965"/>
      <origin xyz="-0.006305773955890981 3.0141314376408408e-5 0.006373188453183354" rpy="0 -0 0"/>
      <inertia ixx="3.848306673536226e-5" ixy="9.375752008419092e-8" ixz="4.870043501240908e-8" iyy="4.53837665921668e-5" iyz="-8.622664755132242e-6" izz="3.7535538604448434e-5"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head_neck_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head_neck_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_CHEST--SIM_ECUB_R_SHOULDER_1" type="revolute">
    <origin xyz="-0.013518676842610352 -0.06704761117227452 0.12351981080941783" rpy="-0.25361232387326116 -0.05203160550449495 0.2752787324138702"/>
    <axis xyz="-0.27144732751428424 -0.9351131265310277 -0.22777135240130472"/>
    <parent link="SIM_ECUB_CHEST"/>
    <child link="SIM_ECUB_R_SHOULDER_1"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_R_SHOULDER_1">
    <inertial>
      <mass value="0.7922822880502421"/>
      <origin xyz="1.7243569867333997e-4 -0.06784040318161476 4.760009452753067e-4" rpy="0 -0 0"/>
      <inertia ixx="4.3422182324316314e-4" ixy="-2.0510369914064142e-4" ixz="1.9458763763805835e-4" iyy="0.0011552753477506248" iyz="4.97645732286129e-5" izz="0.0011466564773150232"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_r_shoulder_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_r_shoulder_1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_R_SHOULDER_1--SIM_ECUB_R_SHOULDER_2" type="revolute">
    <origin xyz="0.007499999999999977 -0.09330000000000004 -3.885780586188048e-16" rpy="-9.325873406851307e-15 9.520162436160717e-15 -4.718447854656912e-15"/>
    <axis xyz="1.0000000000000007 3.700165174258527e-15 9.783840404509184e-15"/>
    <parent link="SIM_ECUB_R_SHOULDER_1"/>
    <child link="SIM_ECUB_R_SHOULDER_2"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_R_SHOULDER_2">
    <inertial>
      <mass value="0.4392272062642899"/>
      <origin xyz="0.005127191267809918 -1.6659290147985614e-4 -0.02041441690188461" rpy="0 -0 0"/>
      <inertia ixx="3.9569579075834417e-4" ixy="5.024923153599911e-5" ixz="-6.506905358695258e-5" iyy="3.2092942321175174e-4" iyz="7.819501283435284e-5" izz="2.674404937949735e-4"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_r_shoulder_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_r_shoulder_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_HEAD_NECK_2--SIM_ECUB_HEAD_NECK_3" type="revolute">
    <origin xyz="8.673617379884035e-19 6.416192023745373e-17 0.009499999999999786" rpy="-1.7364294099408703e-15 -2.980403162017727e-16 1.580860396783509e-15"/>
    <axis xyz="-1 3.264103040198241e-15 4.068108956268403e-15"/>
    <parent link="SIM_ECUB_HEAD_NECK_2"/>
    <child link="SIM_ECUB_HEAD_NECK_3"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_HEAD_NECK_3">
    <inertial>
      <mass value="0.17811376821431446"/>
      <origin xyz="7.590511663558798e-4 -4.091690693859513e-5 0.03603551654922699" rpy="0 -0 0"/>
      <inertia ixx="1.0759851740143372e-4" ixy="-1.762946883518242e-7" ixz="-2.5039696740768445e-7" iyy="9.705486922197742e-5" iyz="-6.130672988865115e-6" izz="3.6264154407241086e-5"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head_neck_3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head_neck_3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_HEAD_NECK_3--SIM_ECUB_HEAD" type="revolute">
    <origin xyz="-1.43982048506075e-16 -4.1419339784385294e-17 0.06914999999999999" rpy="1.2720090288173385e-15 -3.1687776687987536e-15 3.715432083235576e-16"/>
    <axis xyz="1.1543337812533589e-10 8.494511599802698e-16 0.9999999999999998"/>
    <parent link="SIM_ECUB_HEAD_NECK_3"/>
    <child link="SIM_ECUB_HEAD"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_HEAD">
    <inertial>
      <mass value="1.972127893184972"/>
      <origin xyz="-0.01620091407966073 -0.002268813231724585 0.03500744217627949" rpy="0 -0 0"/>
      <inertia ixx="0.009883124315185843" ixy="-1.818674034830039e-4" ixz="-3.994074773445533e-5" iyy="0.007991907355798097" iyz="1.9244538979180473e-4" izz="0.00786000669473206"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_head.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_HEAD--SIM_ECUB_REALSENSE" type="revolute">
    <origin xyz="0.06250999999999993 -3.0000000001105853e-5 0.07639250000000064" rpy="9.166950960511526e-15 2.595819949100105e-4 1.9343583275079285e-15"/>
    <axis xyz="-3.9202157447443984e-15 -1 6.441144250374745e-15"/>
    <parent link="SIM_ECUB_HEAD"/>
    <child link="SIM_ECUB_REALSENSE"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_REALSENSE">
    <inertial>
      <mass value="0.0973042045264468"/>
      <origin xyz="-0.01566548310600984 0.00451999521864419 0.0027456101541037943" rpy="0 -0 0"/>
      <inertia ixx="1.870864157571476e-4" ixy="-4.787553659410962e-5" ixz="-7.183500075449046e-6" iyy="1.3364464963624943e-4" iyz="-2.552445231128513e-5" izz="2.9631966228712615e-4"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_realsense.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_realsense.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
  <joint name="SIM_ECUB_L_SHOULDER_1--SIM_ECUB_L_SHOULDER_2" type="revolute">
    <origin xyz="0.005000000000000011 0.09329999999999987 -3.0531133177191805e-16" rpy="2.6895152771544417e-14 1.790234627208065e-15 -2.3314683517128287e-15"/>
    <axis xyz="1 3.1450536619460376e-15 5.002942504717091e-15"/>
    <parent link="SIM_ECUB_L_SHOULDER_1"/>
    <child link="SIM_ECUB_L_SHOULDER_2"/>
    <limit lower="0" upper="3.141592653589793" effort="1e+9" velocity="1e+9"/>
  </joint>
  <link name="SIM_ECUB_L_SHOULDER_2">
    <inertial>
      <mass value="0.44093319265387243"/>
      <origin xyz="0.007505823346010238 2.3402763110361158e-4 -0.02049745115620874" rpy="0 -0 0"/>
      <inertia ixx="3.9954225516225387e-4" ixy="-5.157553010803303e-5" ixz="6.569228591220418e-5" iyy="3.2267244497365393e-4" iyz="7.978409562239449e-5" izz="2.7204910861740035e-4"/>
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_l_shoulder_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 -0 0"/>
      <geometry>
        <mesh filename="sim_ecub_l_shoulder_2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </collision>
  </link>
</robot>

```

@traversaro do you have any hint why they could be omitted in the exportation?

**Comment by traversaro** [comment]
There is no joint connecting `SIM_ECUB_L_SHOULDER_2` and `SIM_ECUB_L_SHOULDER_3` (and `SIM_ECUB_R_SHOULDER_2` and `SIM_ECUB_R_SHOULDER_3`, so the traversal generated in https://github.com/robotology/idyntree/blob/master/src/model_io/codecs/src/URDFModelExport.cpp#L531 does not reach SIM_ECUB_L_SHOULDER_3 and its children. Probably we could add `Model::isValid` or `Model::isConnected` to ensure that a manually built model is well connected and to early catch problems like this one.

**Comment by Nicogene** [comment]
> Probably we could add Model::isValid or Model::isConnected to ensure that a manually built model is well connected and to early catch problems like this one.

We are already calling `isValid` but probably we have to call also `isConnected`.

> There is no joint connecting SIM_ECUB_L_SHOULDER_2 and SIM_ECUB_L_SHOULDER_3 (and SIM_ECUB_R_SHOULDER_2 and SIM_ECUB_R_SHOULDER_3),

Good catch thanks!

**Comment by traversaro** [comment]
> > Probably we could add Model::isValid or Model::isConnected to ensure that a manually built model is well connected and to early catch problems like this one.
> 
> We are already calling `isValid` but probably we have to call also `isConnected`.

Sorry, those methods do not exists. I meant "add" as in "implement".

**Comment by Nicogene** [comment]
I have understood what is the problem here, right now we always considered that two link are connected by a (revolute) joint, and the assemble has been made on the rotation axis of the joint. 

For `SHOULDER_2` and `SHOULDER_3` unfortunately, the mechanism is different, we should consider a fixed joint between those
since that is the fake joint of the ft sensors.

cc @fiorisi @mfussi66 @pattacini @Mick3Lozzo

**Comment by Nicogene** [comment]
Today I analyzed how was defined the constraints between links for defining the joints and I discovered:

![immagine](https://github.com/icub-tech-iit/creo2urdf/assets/19152494/1a2cdb46-1396-4035-af9c-7f4cf8aa8365)

In the assembly provided by @Mick3Lozzo the revolute joints are defined aligning axis (`ALIGN`) and the fixed joints are defined aligning csys (`CSYS`).
Speaking w/ @fiorisi, this is just a way for defining joints in creo, they can be defined also aligning planes, axis+planes, etc.
For this reason, he suggested changing the upperbody assembly in order to define joints using different techniques, in this way, we can test how `creo2urdf` generalize and handle all these possibilities.
This could be an activity that can be done next sprint/PI.
How the joint has been created could be in-influent if we could access via API the to the Mech constraint object (toolkit for mechanica license), but right now we are forced to assume that joints are defined in a certain way.


>In the assembly provided by @Mick3Lozzo the revolute joints are defined aligning axis (`ALIGN`) and the fixed joints are defined aligning csys (`CSYS`).

In any case, in order to reach the MVP 2.5 I took this assumption for granted, and voila the complete urdf of the upperbody of ergoCub

![immagine](https://github.com/icub-tech-iit/creo2urdf/assets/19152494/6840830f-5bf1-4750-afd8-d34c10cc1b54)

cc @maggia80 @traversaro @pattacini @mfussi66
